File: go/agentdb/project_settings.go
package agentdb

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"gorm.io/gorm"
)

// Spec defaults for the numeric project settings (docs/product/01-session-config.md §5).
const (
	// DefaultMaxConcurrentJobs is the router/scheduler concurrency cap (§8.4).
	DefaultMaxConcurrentJobs = 4
	// DefaultBriefingMaxBytes caps each injected briefing section (§7.4).
	DefaultBriefingMaxBytes = 2048
	// DefaultSnapshotTTLDays is the snapshot reaper horizon; 0 means never (§5).
	DefaultSnapshotTTLDays = 30
)

// ErrInvalidProjectSettings marks a caller mistake (empty project, negative
// budget) as opposed to a store failure, so HTTP handlers can answer 400.
var ErrInvalidProjectSettings = errors.New("invalid project settings")

// ProjectSettings is the per-project configuration row (§5): one row per
// project (the customer string), created lazily on first write. Projects
// themselves stay "a name that exists once something carries it", so reads of
// an unwritten project return the defaults rather than an error.
//
// MCPConfig holds map[string]MCPServerConfig as raw JSON — values are only ever
// ${VAR} references, never secrets (§4.4), so persisting/displaying it is safe.
// Deliberately no gorm `default:` tags on the numeric columns: gorm treats a
// zero value on a defaulted column as "unset" and substitutes the default,
// which would make 0 (= off / never) unwritable. The column DEFAULTs live in
// migration 020 for rows inserted outside this store; normalize() owns the
// in-Go defaulting.
type ProjectSettings struct {
	Project           string  `json:"project" gorm:"primaryKey;type:varchar(255)"`
	BaseImage         string  `json:"base_image" gorm:"type:text"`
	SystemPrompt      string  `json:"system_prompt" gorm:"type:text"`
	MCPConfig         JSONMap `json:"mcp_config" gorm:"type:jsonb"`
	AttentionChannel  JSONMap `json:"attention_channel" gorm:"type:jsonb"`
	MaxConcurrentJobs int     `json:"max_concurrent_jobs"`
	DailyTokensSoft   int64   `json:"daily_tokens_soft"` // 0 = off
	DailyTokensHard   int64   `json:"daily_tokens_hard"` // 0 = off
	BriefingMaxBytes  int     `json:"briefing_max_bytes"`
	SnapshotTTLDays   int     `json:"snapshot_ttl_days"` // 0 = never reap
	UpdatedAt         int64   `json:"updated_at" gorm:"autoUpdateTime"`
}

func (ProjectSettings) TableName() string { return "project_settings" }

// DefaultProjectSettings returns the settings a project has before anything has
// ever been written for it. Not persisted — GetProjectSettings hands this back
// for an unknown project (lazy creation happens on the first write).
func DefaultProjectSettings(project string) *ProjectSettings {
	return &ProjectSettings{
		Project:           project,
		MCPConfig:         JSONMap{},
		AttentionChannel:  JSONMap{},
		MaxConcurrentJobs: DefaultMaxConcurrentJobs,
		BriefingMaxBytes:  DefaultBriefingMaxBytes,
		SnapshotTTLDays:   DefaultSnapshotTTLDays,
	}
}

// normalize validates a whole-object write and fills the "unset" numerics.
//
// Zero is a *meaningful* value for the two settings the spec says so about —
// daily_tokens_soft/hard (0 = off) and snapshot_ttl_days (0 = never) — so those
// are kept as written. For max_concurrent_jobs and briefing_max_bytes zero has
// no useful meaning (it would deadlock the router / delete every briefing), so
// it is read as "unset" and the spec default applies.
func (ps *ProjectSettings) normalize() error {
	if ps.Project == "" {
		return fmt.Errorf("%w: project is required", ErrInvalidProjectSettings)
	}
	for _, f := range []struct {
		name string
		v    int64
	}{
		{"max_concurrent_jobs", int64(ps.MaxConcurrentJobs)},
		{"daily_tokens_soft", ps.DailyTokensSoft},
		{"daily_tokens_hard", ps.DailyTokensHard},
		{"briefing_max_bytes", int64(ps.BriefingMaxBytes)},
		{"snapshot_ttl_days", int64(ps.SnapshotTTLDays)},
	} {
		if f.v < 0 {
			return fmt.Errorf("%w: %s must not be negative (got %d)", ErrInvalidProjectSettings, f.name, f.v)
		}
	}
	if ps.MaxConcurrentJobs == 0 {
		ps.MaxConcurrentJobs = DefaultMaxConcurrentJobs
	}
	if ps.BriefingMaxBytes == 0 {
		ps.BriefingMaxBytes = DefaultBriefingMaxBytes
	}
	if ps.MCPConfig == nil {
		ps.MCPConfig = JSONMap{}
	}
	if ps.AttentionChannel == nil {
		ps.AttentionChannel = JSONMap{}
	}
	return nil
}

// GetProjectSettings returns the settings row for project, or the defaults when
// the project has never been written. Project scoping is enforced here and ONLY
// here: the caller passes the project it is authorized for and can reach no other.
func (s *Store) GetProjectSettings(ctx context.Context, project string) (*ProjectSettings, error) {
	if project == "" {
		return nil, fmt.Errorf("%w: project is required", ErrInvalidProjectSettings)
	}
	var ps ProjectSettings
	err := s.gdb.WithContext(ctx).Where("project = ?", project).First(&ps).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return DefaultProjectSettings(project), nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get project settings: %w", err)
	}
	return &ps, nil
}

// PutProjectSettings writes the whole settings object for ps.Project (no patch
// semantics — §5: "PUT is whole-object"), creating the row on first write, and
// returns the stored state read back from the row that was written.
//
// The write appends a `project_settings_put` config event in the same
// transaction (§15.3/§15.4). It is one action whether the row was created or
// replaced: §5's settings row has no create/update distinction — it is created
// lazily and always written whole. cw carries the who/why; a human/API edit
// passes the zero value.
func (s *Store) PutProjectSettings(ctx context.Context, ps *ProjectSettings, cw ConfigWrite) (*ProjectSettings, error) {
	if ps == nil {
		return nil, fmt.Errorf("%w: settings are required", ErrInvalidProjectSettings)
	}
	next := *ps
	if err := next.normalize(); err != nil {
		return nil, err
	}

	var existing ProjectSettings
	err := s.gdb.WithContext(ctx).Where("project = ?", next.Project).First(&existing).Error
	switch {
	case err == nil:
		// Whole-object replace: every field is written, zero values included.
		existing.BaseImage = next.BaseImage
		existing.SystemPrompt = next.SystemPrompt
		existing.MCPConfig = next.MCPConfig
		existing.AttentionChannel = next.AttentionChannel
		existing.MaxConcurrentJobs = next.MaxConcurrentJobs
		existing.DailyTokensSoft = next.DailyTokensSoft
		existing.DailyTokensHard = next.DailyTokensHard
		existing.BriefingMaxBytes = next.BriefingMaxBytes
		existing.SnapshotTTLDays = next.SnapshotTTLDays
		if _, err := s.WithConfigEvent(ctx, ConfigChange{
			Project: existing.Project,
			Action:  ActionProjectSettingsPut,
			Payload: &existing,
			Write:   cw,
		}, func(tx *gorm.DB) error {
			return tx.Save(&existing).Error
		}); err != nil {
			return nil, fmt.Errorf("failed to update project settings: %w", err)
		}
		return &existing, nil
	case errors.Is(err, gorm.ErrRecordNotFound):
		if _, err := s.WithConfigEvent(ctx, ConfigChange{
			Project: next.Project,
			Action:  ActionProjectSettingsPut,
			Payload: &next,
			Write:   cw,
		}, func(tx *gorm.DB) error {
			return tx.Create(&next).Error
		}); err != nil {
			return nil, fmt.Errorf("failed to create project settings: %w", err)
		}
		return &next, nil
	default:
		return nil, fmt.Errorf("failed to read project settings: %w", err)
	}
}

// SetProjectPrompt replaces the project-level system prompt wholesale (P4) and
// appends a `project_prompt_write` config event. It is the ONLY path that writes
// that action.
//
// It is a NARROW write rather than a read-modify-write through
// PutProjectSettings for the same reason SetWorkerPrompt is: PutProjectSettings
// is whole-object (§5), so rewriting the prompt through it would silently
// rewrite the budgets and the attention channel with whatever the caller last
// read — and a prompt rewrite is precisely the mutation that must change nothing
// else. It also lets the two actions stay distinct in the log: `project_prompt_write`
// carries a rationale (§15.5) and folds as its own entity (§15.6).
//
// The payload is `{project, system_prompt}` — NOT the whole settings row. That
// shape is pinned in three places already (the fold's EntityProjectPrompt
// singleton, its tests, and the changelog UI, which accepts `system_prompt`
// first — the F1 finding of 2026-07-25), and it is what §15.3 asks for: "the new
// project system prompt".
//
// The superseded prompt is returned alongside the stored row for the automatic
// `kind=prompt-revision` memory (§9).
func (s *Store) SetProjectPrompt(ctx context.Context, project, prompt string, cw ConfigWrite) (*ProjectSettings, string, error) {
	if strings.TrimSpace(project) == "" {
		return nil, "", fmt.Errorf("%w: project is required", ErrInvalidProjectSettings)
	}
	if strings.TrimSpace(prompt) == "" {
		return nil, "", fmt.Errorf("%w: system_prompt must not be blank (§9 validates a non-empty prompt)", ErrInvalidProjectSettings)
	}

	change := ConfigChange{
		Project: project,
		Action:  ActionProjectPromptWrite,
		Payload: JSONMap{"project": project, "system_prompt": prompt},
		Write:   cw,
	}

	var existing ProjectSettings
	err := s.gdb.WithContext(ctx).Where("project = ?", project).First(&existing).Error
	switch {
	case err == nil:
		previous := existing.SystemPrompt
		now := time.Now().Unix()
		if _, err := s.WithConfigEvent(ctx, change, func(tx *gorm.DB) error {
			return tx.Model(&ProjectSettings{}).
				Where("project = ?", project).
				Updates(map[string]any{"system_prompt": prompt, "updated_at": now}).Error
		}); err != nil {
			return nil, "", fmt.Errorf("failed to write project prompt: %w", err)
		}
		stored, err := s.GetProjectSettings(ctx, project)
		if err != nil {
			return nil, previous, err
		}
		return stored, previous, nil

	case errors.Is(err, gorm.ErrRecordNotFound):
		// The settings row is created lazily on first write (§5), so the first
		// project_prompt_write for a project creates it — with the spec defaults,
		// not with zeroes, because a row of zeroes would read as "concurrency 0".
		next := DefaultProjectSettings(project)
		next.SystemPrompt = prompt
		if _, err := s.WithConfigEvent(ctx, change, func(tx *gorm.DB) error {
			return tx.Create(next).Error
		}); err != nil {
			return nil, "", fmt.Errorf("failed to create project settings: %w", err)
		}
		stored, err := s.GetProjectSettings(ctx, project)
		if err != nil {
			return nil, "", err
		}
		// Nothing was superseded: this project had never had a prompt.
		return stored, "", nil

	default:
		return nil, "", fmt.Errorf("failed to read project settings: %w", err)
	}
}

File: web/src/projectSettings.ts
// Project settings — the browser-side mirror of `GET/PUT /agent/project-settings`
// (spec docs/product/01-session-config.md §5, engine: go/agentdb/project_settings.go).
//
// Everything here is pure: no React, no window, no fetch. The hook that talks to
// the API is useProjectSettings.ts; the page is components/ProjectSettingsPage.tsx.
//
// Two things this module exists to get right:
//
//  1. **JSON before save.** `mcp_config` and `attention_channel` are free-form
//     JSON objects. The editor must refuse to PUT text that is not a JSON object
//     — a malformed blob would either 400 or, worse, round-trip as a string and
//     silently disable every project MCP server. `parseJsonObject` is the gate.
//  2. **Zero is not "empty".** Four of the numeric settings treat 0 as a real,
//     meaningful value ("off" / "never"); two treat it as "unset" and the server
//     substitutes a default. Getting that backwards in the UI is how a human
//     accidentally uncaps their token spend. `PROJECT_SETTING_NUMERICS` carries
//     the semantics as data so the form renders the truth rather than a guess,
//     and it mirrors `ProjectSettings.normalize()` in the engine exactly.

/** Default endpoint for the project-settings routes (GET and PUT share it). */
export const PROJECT_SETTINGS_ENDPOINT = '/agent/project-settings'

/** Engine defaults — go/agentdb/project_settings.go. Kept in step by eye; the
 *  server is authoritative and its response overwrites these on first load. */
export const DEFAULT_MAX_CONCURRENT_JOBS = 4
export const DEFAULT_BRIEFING_MAX_BYTES = 2048
export const DEFAULT_SNAPSHOT_TTL_DAYS = 30

/** One project's settings row. Field names are the wire names (snake_case). */
export interface ProjectSettings {
  project: string
  base_image: string
  system_prompt: string
  mcp_config: Record<string, unknown>
  attention_channel: Record<string, unknown>
  max_concurrent_jobs: number
  daily_tokens_soft: number
  daily_tokens_hard: number
  briefing_max_bytes: number
  snapshot_ttl_days: number
  updated_at: number
}

/** The settings a project has before anything has been written for it — the
 *  same shape the server hands back for an unwritten project. */
export function defaultProjectSettings(project = ''): ProjectSettings {
  return {
    project,
    base_image: '',
    system_prompt: '',
    mcp_config: {},
    attention_channel: {},
    max_concurrent_jobs: DEFAULT_MAX_CONCURRENT_JOBS,
    daily_tokens_soft: 0,
    daily_tokens_hard: 0,
    briefing_max_bytes: DEFAULT_BRIEFING_MAX_BYTES,
    snapshot_ttl_days: DEFAULT_SNAPSHOT_TTL_DAYS,
    updated_at: 0,
  }
}

/** Fill in anything the server omitted, so the form never binds `undefined`
 *  to a controlled input (React would flip it to uncontrolled mid-edit). */
export function coerceProjectSettings(raw: unknown, project = ''): ProjectSettings {
  const base = defaultProjectSettings(project)
  if (!raw || typeof raw !== 'object') return base
  const r = raw as Record<string, unknown>
  const str = (k: keyof ProjectSettings, d: string) =>
    typeof r[k] === 'string' ? (r[k] as string) : d
  const num = (k: keyof ProjectSettings, d: number) =>
    typeof r[k] === 'number' && Number.isFinite(r[k] as number) ? (r[k] as number) : d
  const obj = (k: keyof ProjectSettings) =>
    r[k] && typeof r[k] === 'object' && !Array.isArray(r[k])
      ? (r[k] as Record<string, unknown>)
      : {}
  return {
    project: str('project', base.project),
    base_image: str('base_image', base.base_image),
    system_prompt: str('system_prompt', base.system_prompt),
    mcp_config: obj('mcp_config'),
    attention_channel: obj('attention_channel'),
    max_concurrent_jobs: num('max_concurrent_jobs', base.max_concurrent_jobs),
    daily_tokens_soft: num('daily_tokens_soft', base.daily_tokens_soft),
    daily_tokens_hard: num('daily_tokens_hard', base.daily_tokens_hard),
    briefing_max_bytes: num('briefing_max_bytes', base.briefing_max_bytes),
    snapshot_ttl_days: num('snapshot_ttl_days', base.snapshot_ttl_days),
    updated_at: num('updated_at', 0),
  }
}

// ---------------------------------------------------------------------------
// JSON editing
// ---------------------------------------------------------------------------

/** Result of parsing an editor's JSON text. Discriminated so callers cannot
 *  read `value` without having checked `ok` first. */
export type JsonParseResult =
  | { ok: true; value: Record<string, unknown> }
  | { ok: false; error: string }

/**
 * Parse the text of a JSON-object editor. Empty (or whitespace-only) text is
 * the empty object — "I cleared the field" must not read as a syntax error.
 * Arrays, strings and numbers are rejected: both settings are name-keyed maps,
 * and a JSON array would be accepted by `JSON.parse` and then silently mean
 * nothing to the server.
 */
export function parseJsonObject(text: string): JsonParseResult {
  const trimmed = text.trim()
  if (trimmed === '') return { ok: true, value: {} }
  let parsed: unknown
  try {
    parsed = JSON.parse(trimmed)
  } catch (err) {
    return { ok: false, error: err instanceof Error ? err.message : 'invalid JSON' }
  }
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return { ok: false, error: 'must be a JSON object, e.g. {"name": {...}}' }
  }
  return { ok: true, value: parsed as Record<string, unknown> }
}

/** Pretty-print an object for a JSON editor. An empty object renders as `{}`
 *  rather than `{\n}` so a fresh project shows a one-line, obviously-empty box. */
export function formatJsonObject(value: unknown): string {
  if (!value || typeof value !== 'object') return '{}'
  if (!Array.isArray(value) && Object.keys(value as object).length === 0) return '{}'
  return JSON.stringify(value, null, 2)
}

// ---------------------------------------------------------------------------
// Numeric settings — the §5 budget/cap fields and their zero semantics
// ---------------------------------------------------------------------------

/**
 * How a numeric setting behaves at zero.
 *   'meaningful' — 0 is a real setting the server stores and honours.
 *   'unset'      — 0 is read as "not supplied" and the server substitutes
 *                  `serverDefault` (ProjectSettings.normalize in the engine).
 */
export type ZeroSemantics = 'meaningful' | 'unset'

export interface NumericSettingSpec {
  /** Wire field name. */
  key:
    | 'max_concurrent_jobs'
    | 'daily_tokens_soft'
    | 'daily_tokens_hard'
    | 'briefing_max_bytes'
    | 'snapshot_ttl_days'
  /** Field label for the form. */
  label: string
  /** What the number counts, shown as the input's suffix. */
  unit: string
  /** One line explaining the setting when the value is non-zero. */
  help: string
  zeroSemantics: ZeroSemantics
  /** The sentence shown *instead of* `help` when the value is 0. */
  zeroHelp: string
  /** Only meaningful when zeroSemantics === 'unset'. */
  serverDefault: number
}

/**
 * The five numeric project settings, in the order the form shows them. The four
 * §5 budget/cap fields plus `max_concurrent_jobs`, which shares their shape.
 *
 * The `zeroSemantics`/`zeroHelp` pairs are a direct transcription of the
 * engine's `normalize()`: daily_tokens_soft/hard and snapshot_ttl_days keep a
 * written 0; max_concurrent_jobs and briefing_max_bytes read 0 as unset because
 * a literal zero there would deadlock the router / delete every briefing.
 */
export const PROJECT_SETTING_NUMERICS: readonly NumericSettingSpec[] = [
  {
    key: 'daily_tokens_soft',
    label: 'Daily token budget — soft',
    unit: 'tokens/day',
    help: 'Crossing this sends one notification to the attention channel per day. Nothing stops.',
    zeroSemantics: 'meaningful',
    zeroHelp: '0 means off — no soft-budget notification is ever sent.',
    serverDefault: 0,
  },
  {
    key: 'daily_tokens_hard',
    label: 'Daily token budget — hard',
    unit: 'tokens/day',
    help: 'Crossing this stops creation of non-interactive jobs until midnight. Interactive chat is never blocked.',
    zeroSemantics: 'meaningful',
    zeroHelp: '0 means off — jobs are never stopped for budget.',
    serverDefault: 0,
  },
  {
    key: 'briefing_max_bytes',
    label: 'Briefing section cap',
    unit: 'bytes/section',
    help: 'Byte cap applied to each injected briefing section independently, at composition time.',
    zeroSemantics: 'unset',
    zeroHelp: `0 is read as "not set" — the server applies its default of ${DEFAULT_BRIEFING_MAX_BYTES} bytes. There is no way to switch briefing off here.`,
    serverDefault: DEFAULT_BRIEFING_MAX_BYTES,
  },
  {
    key: 'snapshot_ttl_days',
    label: 'Snapshot TTL',
    unit: 'days',
    help: 'The reaper deletes a snapshot image this many days after it was created.',
    zeroSemantics: 'meaningful',
    zeroHelp: '0 means never — snapshots are kept forever and no reaper touches them.',
    serverDefault: 0,
  },
  {
    key: 'max_concurrent_jobs',
    label: 'Max concurrent jobs',
    unit: 'jobs',
    help: 'Router/scheduler concurrency cap for the whole project.',
    zeroSemantics: 'unset',
    zeroHelp: `0 is read as "not set" — the server applies its default of ${DEFAULT_MAX_CONCURRENT_JOBS}. Use 1 to serialise the project.`,
    serverDefault: DEFAULT_MAX_CONCURRENT_JOBS,
  },
]

/**
 * The helper line to show under a numeric field for the value currently typed —
 * the whole point of the spec's "0 means…" wording being legible to a human.
 */
export function describeNumericSetting(spec: NumericSettingSpec, value: number): string {
  if (value === 0) return spec.zeroHelp
  return spec.help
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

/** Field name → human-readable problem. Empty object means "safe to save". */
export type FieldErrors = Record<string, string>

/**
 * Client-side validation. Mirrors the engine's rules (negatives rejected,
 * everything else permitted) rather than inventing stricter ones — the server
 * stays authoritative, this just avoids a pointless round-trip and gives the
 * human the error next to the field instead of in a toast.
 */
export function validateProjectSettings(s: ProjectSettings): FieldErrors {
  const errors: FieldErrors = {}
  for (const spec of PROJECT_SETTING_NUMERICS) {
    const v = s[spec.key]
    if (!Number.isFinite(v) || !Number.isInteger(v)) {
      errors[spec.key] = 'must be a whole number'
    } else if (v < 0) {
      errors[spec.key] = 'must not be negative'
    }
  }
  return errors
}

/**
 * The PUT body. `project` and `updated_at` are dropped: the server takes the
 * project from the JWT and stamps the timestamp itself, and sending our own
 * values invites the illusion that either is client-settable.
 *
 * `rationale` is the operator's one-line reason, threaded into the config event
 * (design B3 / K2), and omitted when empty so an absent reason reads as absent
 * — the same contract `scheduleBody` has.
 */
export function projectSettingsBody(
  s: ProjectSettings,
  rationale = '',
): Omit<ProjectSettings, 'project' | 'updated_at'> & { rationale?: string } {
  const { project: _project, updated_at: _updatedAt, ...body } = s
  return rationale.trim() === '' ? body : { ...body, rationale: rationale.trim() }
}
